{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "24aed7d8-b802-4e92-bb23-6dc2d5341f3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== CpG Methylation Matrix ===\n",
      "                P102d     P105d     P111d     P113d     P117d     P118d  \\\n",
      "Gene_Symbol                                                               \n",
      "A1BG         0.582906  0.456747  0.535035  0.638074  0.615732  0.505413   \n",
      "A1BG-AS1     0.500771  0.376418  0.397345  0.521447  0.550180  0.339344   \n",
      "A2M-AS1      0.666854  0.713421  0.561839  0.573033  0.694567  0.603831   \n",
      "A2ML1        0.808487  0.788236  0.829482  0.878738  0.859099  0.831974   \n",
      "A2ML1-AS1    0.834131  0.876004  0.814006  0.841255  0.826644  0.819913   \n",
      "\n",
      "                P123d     P127d     P128d     P131d  ...      P66d      P68d  \\\n",
      "Gene_Symbol                                          ...                       \n",
      "A1BG         0.554163  0.541218  0.586789  0.689665  ...  0.646065  0.626514   \n",
      "A1BG-AS1     0.444468  0.453535  0.412788  0.600901  ...  0.563106  0.554267   \n",
      "A2M-AS1      0.629768  0.699405  0.587279  0.601171  ...  0.729268  0.492997   \n",
      "A2ML1        0.850209  0.821716  0.862571  0.878696  ...  0.814136  0.827510   \n",
      "A2ML1-AS1    0.850567  0.908111  0.844114  0.888498  ...  0.824172  0.842005   \n",
      "\n",
      "                 P73d      P78d      P81d      P86d      P87d      P96d  \\\n",
      "Gene_Symbol                                                               \n",
      "A1BG         0.539323  0.595239  0.547809  0.518117  0.558065  0.648699   \n",
      "A1BG-AS1     0.406840  0.456769  0.405290  0.328443  0.424231  0.485457   \n",
      "A2M-AS1      0.668561  0.654811  0.661394  0.627671  0.703013  0.538907   \n",
      "A2ML1        0.858536  0.846348  0.859003  0.899886  0.874532  0.852631   \n",
      "A2ML1-AS1    0.834358  0.901724  0.902159  0.859765  0.813304  0.834404   \n",
      "\n",
      "                 P98d      P99d  \n",
      "Gene_Symbol                      \n",
      "A1BG         0.557049  0.495684  \n",
      "A1BG-AS1     0.490016  0.350884  \n",
      "A2M-AS1      0.576969  0.702448  \n",
      "A2ML1        0.810195  0.790069  \n",
      "A2ML1-AS1    0.831451  0.833583  \n",
      "\n",
      "[5 rows x 50 columns]\n",
      "\n",
      "=== miRNA Expression Matrix ===\n",
      "                       P102s        P105s        P111s        P113s  \\\n",
      "sRNA                                                                  \n",
      "hsa-let-7a-2-3p     0.000000     0.000000     0.000000     0.000000   \n",
      "hsa-let-7a-3p      30.341054    44.325241    42.349801    20.402847   \n",
      "hsa-let-7a-5p    3044.004813  3761.834535  5573.501486  5559.970235   \n",
      "hsa-let-7b-3p       0.642819     0.675690     2.190507     1.943128   \n",
      "hsa-let-7b-5p     153.248035   347.304478   355.592291   298.173042   \n",
      "\n",
      "                       P117s        P118s        P123s        P127s  \\\n",
      "sRNA                                                                  \n",
      "hsa-let-7a-2-3p     0.000000     0.000000     0.000000     0.000000   \n",
      "hsa-let-7a-3p      27.398130    20.295304    32.510492    68.927690   \n",
      "hsa-let-7a-5p    5398.833298  9218.386371  4292.027489  9415.522432   \n",
      "hsa-let-7b-3p       2.293797     1.295445     1.927500     3.759692   \n",
      "hsa-let-7b-5p     276.912259   369.979082   268.179437  1133.199071   \n",
      "\n",
      "                       P128s         P131s  ...         P66s         P68s  \\\n",
      "sRNA                                        ...                             \n",
      "hsa-let-7a-2-3p     0.161335      0.000000  ...     0.117212     0.000000   \n",
      "hsa-let-7a-3p      25.490969     11.797041  ...    71.030727    28.065760   \n",
      "hsa-let-7a-5p    5063.910136  12389.120930  ...  4105.482271  6028.647332   \n",
      "hsa-let-7b-3p       0.887344      1.441861  ...     0.586062     0.122025   \n",
      "hsa-let-7b-5p     165.852636    885.433434  ...   183.789077    89.810433   \n",
      "\n",
      "                        P73s         P78s         P81s         P86s  \\\n",
      "sRNA                                                                  \n",
      "hsa-let-7a-2-3p     0.000000     0.000000     0.000000     0.000000   \n",
      "hsa-let-7a-3p      31.102083    48.437008    29.236495    30.366046   \n",
      "hsa-let-7a-5p    7704.846788  6826.216721  4689.799580  9455.141816   \n",
      "hsa-let-7b-3p       1.249637     0.136059     0.000000     0.342474   \n",
      "hsa-let-7b-5p     279.085654   207.626053    86.501365   273.636884   \n",
      "\n",
      "                         P87s         P96s         P98s         P99s  \n",
      "sRNA                                                                  \n",
      "hsa-let-7a-2-3p      0.000000     0.000000     0.000000     0.000000  \n",
      "hsa-let-7a-3p       42.718215    32.075481    44.376706    36.064500  \n",
      "hsa-let-7a-5p    13303.989320  3902.711218  5853.325805  6585.176028  \n",
      "hsa-let-7b-3p        1.478139     1.399657     1.721510     0.630498  \n",
      "hsa-let-7b-5p      795.090923    74.531753   286.822720   349.422136  \n",
      "\n",
      "[5 rows x 50 columns]\n",
      "\n",
      "=== mRNA Expression Matrix ===\n",
      "                         P26m          P31m          P33m          P37m  \\\n",
      "ENSG00000213218  13031.394620  54628.551700  64073.976140  125395.84930   \n",
      "ENSG00000136488  19798.206670  55595.539630  61209.326870  104521.52130   \n",
      "ENSG00000156508   8237.236546   8638.347144   4620.087343    3563.53485   \n",
      "ENSG00000105825   5160.342163  28885.268020   9391.500795   16069.42804   \n",
      "ENSG00000198804  30601.842590  15056.787930  16979.119770   14148.44793   \n",
      "\n",
      "                         P42m          P51m          P52m           P55m  \\\n",
      "ENSG00000213218  84797.883570  85677.974950  91199.705530  108029.726600   \n",
      "ENSG00000136488  65546.348340  92972.924610  76486.532730   85963.344440   \n",
      "ENSG00000156508   6642.815691   4994.102332   8710.949269    3904.417469   \n",
      "ENSG00000105825  15801.440500  26438.101010  21813.678480   15999.395380   \n",
      "ENSG00000198804  27128.272270   9553.149666  14749.257490    9557.124608   \n",
      "\n",
      "                         P59m          P60m  ...         P244m        P246m  \\\n",
      "ENSG00000213218  16549.916290  77047.586780  ...   1121.206747  60438.54866   \n",
      "ENSG00000136488  28180.767680  67524.358710  ...   2692.913885  65707.68948   \n",
      "ENSG00000156508   7456.400737   5869.733478  ...   8092.825195   4042.54325   \n",
      "ENSG00000105825   3833.178066  16427.019160  ...    828.694931  10246.55483   \n",
      "ENSG00000198804  18783.307280  16010.269930  ...  28025.505370  13209.18256   \n",
      "\n",
      "                        P257m         P261m        P262m         P266m  \\\n",
      "ENSG00000213218  27710.009710  72030.733070  36878.00691  52133.667430   \n",
      "ENSG00000136488  36270.405000  64821.960230  70921.39051  80074.921010   \n",
      "ENSG00000156508   6190.212424   5961.165682   7195.01574   7694.817784   \n",
      "ENSG00000105825   3408.645647  13525.408370  13353.50174  14476.536430   \n",
      "ENSG00000198804  21412.561590  15235.325870  13311.41575  21365.179250   \n",
      "\n",
      "                         P73m         P118m          P87m          P98m  \n",
      "ENSG00000213218   9106.370742   2673.859121  89397.855650  54363.028640  \n",
      "ENSG00000136488  13715.233160   4079.582223  73964.683150  52007.413280  \n",
      "ENSG00000156508   6135.143708   8604.913296   4497.059161   5309.093377  \n",
      "ENSG00000105825   1553.535276   1502.895272  21229.753090  13559.867790  \n",
      "ENSG00000198804  17887.631230  21326.636490   7208.392151  16884.438390  \n",
      "\n",
      "[5 rows x 50 columns]\n",
      "\n",
      "=== Covariate Matrix ===\n",
      "      DNA_ID mRNA_ID sRNA_ID   age        BMI     BMI_cat B12_status  \\\n",
      "NTUID                                                                  \n",
      "102    P102d   P102m   P102s  36.5  20.703735     healthy         NB   \n",
      "105    P105d   P105m   P105s  33.5  38.404033      obese2         NB   \n",
      "111    P111d   P111m   P111s  32.6  22.519433     healthy         NB   \n",
      "113    P113d   P113m   P113s  31.5  25.970116  overweight         NB   \n",
      "117    P117d   P117m   P117s  37.0  26.370238  overweight         LB   \n",
      "\n",
      "        B12_mol  B12supplem    ethnicity  parity  V1_se_EmploymentStatus  \\\n",
      "NTUID                                                                      \n",
      "102    292.6170         1.0  South Asian       1                     NaN   \n",
      "105    362.2104         1.0    Caucasian       1                     1.0   \n",
      "111    312.3954         1.0        Other       1                     1.0   \n",
      "113    469.2204         1.0        Other       1                     1.0   \n",
      "117    138.3012         2.0  South Asian       1                     5.0   \n",
      "\n",
      "       V1_se_HouseholdIncome  v1p_MultivitTab  v1p_FolicAcid  smoking  \\\n",
      "NTUID                                                                   \n",
      "102                       15              1.0            1.0      2.0   \n",
      "105                        9              1.0            1.0      NaN   \n",
      "111                        9              1.0            1.0      0.0   \n",
      "113                        7              1.0            2.0      0.0   \n",
      "117                       15              2.0            2.0      0.0   \n",
      "\n",
      "       v3n_Gender  batch_miRNA  batch_mRNA batch_DNA  \n",
      "NTUID                                                 \n",
      "102             1            3           3         C  \n",
      "105             1            4           3         C  \n",
      "111             2            2           1         A  \n",
      "113             1            4           1         A  \n",
      "117             2            2           1         A  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "file_paths = {\n",
    "    \"CpG Methylation Matrix\": \"/Users/heweilin/Desktop/P056_Code_2/Processed_Data/1_PD_PromoterRegion_CpGs.csv\",\n",
    "    \"miRNA Expression Matrix\": \"/Users/heweilin/Desktop/P056_Code_2/Raw_Data/6miRNA_TPM.csv\",\n",
    "    \"mRNA Expression Matrix\": \"/Users/heweilin/Desktop/P056_Code_2/Raw_Data/5mRNA_TPM.csv\",\n",
    "    \"Covariate Matrix\": \"/Users/heweilin/Desktop/P056_Code_2/Raw_Data/7Clinical_data50.csv\"\n",
    "}\n",
    "\n",
    "for name, path in file_paths.items():\n",
    "    print(f\"\\n=== {name} ===\")\n",
    "    try:\n",
    "        df = pd.read_csv(path, index_col=0)\n",
    "        print(df.head())\n",
    "    except Exception as e:\n",
    "        print(f\"Error reading {name}: {e}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b2197221-60ba-41cd-b5d6-023a79bdda95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully generated 282 triplets.\n",
      "Saved to: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_candidate_triplets.csv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# === Step 1: Load input files ===\n",
    "# CpG–mRNA regulation data (classical negative regulation)\n",
    "path_dna = \"/Users/heweilin/Desktop/P056_Code_2/Raw_Data/9Integrated_DNA_mRNA.xlsx\"\n",
    "df_dna_hypo = pd.read_excel(path_dna, sheet_name=\"Hypo_Up\")      # CpG ↓ → mRNA ↑\n",
    "df_dna_hyper = pd.read_excel(path_dna, sheet_name=\"Hyper_Down\")  # CpG ↑ → mRNA ↓\n",
    "\n",
    "# miRNA–mRNA regulation data (classical negative regulation)\n",
    "path_mirna = \"/Users/heweilin/Desktop/P056_Code_2/Raw_Data/10Integrated_miRNA_mRNA.xlsx\"\n",
    "df_mirna_up = pd.read_excel(path_mirna, sheet_name=\"mirUP_mrnaDown\")    # miRNA ↑ → mRNA ↓\n",
    "df_mirna_down = pd.read_excel(path_mirna, sheet_name=\"mirDown_mrnaUP\")  # miRNA ↓ → mRNA ↑\n",
    "\n",
    "# Filtered significant CpG–gene pairs from promoter region\n",
    "path_cpg_genes = \"/Users/heweilin/Desktop/P056_Code_2/Processed_Data/1_PD_PromoterRegion_SignificantCpGs.csv\"\n",
    "df_cpg = pd.read_csv(path_cpg_genes)\n",
    "selected_genes = df_cpg[\"Gene_Symbol\"].dropna().unique()\n",
    "\n",
    "# === Step 2: Prepare CpG–Gene and miRNA–Gene pairs ===\n",
    "# Combine CpG–Gene regulation pairs\n",
    "df_dna_all = pd.concat([\n",
    "    df_dna_hypo[[\"name\", \"SYMBOL\"]],\n",
    "    df_dna_hyper[[\"name\", \"SYMBOL\"]]\n",
    "]).rename(columns={\"name\": \"CpG\", \"SYMBOL\": \"Gene\"})\n",
    "\n",
    "# Keep only CpG–Gene pairs where Gene is among significant CpG genes\n",
    "df_dna_all = df_dna_all[df_dna_all[\"Gene\"].isin(selected_genes)]\n",
    "\n",
    "# Combine miRNA–Gene regulation pairs\n",
    "df_mirna_all = pd.concat([\n",
    "    df_mirna_up[[\"mirna\", \"Target.Gene\"]],\n",
    "    df_mirna_down[[\"mirna\", \"Target.Gene\"]]\n",
    "]).rename(columns={\"Target.Gene\": \"Gene\"})\n",
    "\n",
    "# === Step 3: Construct CpG–miRNA–mRNA triplets ===\n",
    "# Only retain genes regulated by both CpG and miRNA\n",
    "shared_genes = set(df_dna_all[\"Gene\"]) & set(df_mirna_all[\"Gene\"])\n",
    "\n",
    "triplets = []\n",
    "for gene in shared_genes:\n",
    "    # In the original data, CpG column actually stores the gene name (not CpG ID)\n",
    "    # So we explicitly rename it as \"CpG_<Gene>\" to distinguish it\n",
    "    cpg_label = f\"CpG_{gene}\"\n",
    "    mirnas = df_mirna_all[df_mirna_all[\"Gene\"] == gene][\"mirna\"].unique()\n",
    "    for mir in mirnas:\n",
    "        triplets.append([cpg_label, mir, gene])\n",
    "\n",
    "# === Step 4: Save the output triplet table ===\n",
    "df_triplets = pd.DataFrame(triplets, columns=[\"CpG\", \"miRNA\", \"mRNA\"])\n",
    "output_path = \"/Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_candidate_triplets.csv\"\n",
    "df_triplets.to_csv(output_path, index=False)\n",
    "\n",
    "print(f\"Successfully generated {len(df_triplets)} triplets.\")\n",
    "print(f\"Saved to: {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f9716514-4ad8-410e-a4be-7fc19cd03460",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DATA FILES PREVIEW REPORT\n",
      "================================================================================\n",
      "\n",
      "================================================================================\n",
      "File: mRNA Expression Data\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/5mRNA_TPM.csv\n",
      "================================================================================\n",
      "Dimensions: 58735 rows x 51 columns\n",
      "Column names: ['Unnamed: 0', 'P26m', 'P31m', 'P33m', 'P37m', 'P42m', 'P51m', 'P52m', 'P55m', 'P59m']\n",
      "              ... (and 41 more columns)\n",
      "Data types:\n",
      "    float64: 50 columns\n",
      "    object: 1 columns\n",
      "\n",
      "First 5 rows:\n",
      "        Unnamed: 0          P26m          P31m          P33m          P37m          P42m          P51m          P52m           P55m          P59m\n",
      "0  ENSG00000213218  13031.394620  54628.551700  64073.976140  125395.84930  84797.883570  85677.974950  91199.705530  108029.726600  16549.916290\n",
      "1  ENSG00000136488  19798.206670  55595.539630  61209.326870  104521.52130  65546.348340  92972.924610  76486.532730   85963.344440  28180.767680\n",
      "2  ENSG00000156508   8237.236546   8638.347144   4620.087343    3563.53485   6642.815691   4994.102332   8710.949269    3904.417469   7456.400737\n",
      "3  ENSG00000105825   5160.342163  28885.268020   9391.500795   16069.42804  15801.440500  26438.101010  21813.678480   15999.395380   3833.178066\n",
      "4  ENSG00000198804  30601.842590  15056.787930  16979.119770   14148.44793  27128.272270   9553.149666  14749.257490    9557.124608  18783.307280\n",
      "\n",
      "No missing values found.\n",
      "\n",
      "Numeric columns summary:\n",
      "               P26m          P31m          P33m          P37m          P42m          P51m          P52m          P55m          P59m          P60m          P64m          P66m          P68m          P78m          P81m          P86m          P96m          P99m         P102m         P105m         P111m         P113m         P117m         P123m         P127m         P128m         P131m         P140m         P146m         P153m         P163m         P185m         P191m         P197m         P204m         P207m         P211m         P219m         P226m         P241m         P244m         P246m         P257m         P261m         P262m         P266m          P73m         P118m          P87m          P98m\n",
      "count  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000  58735.000000\n",
      "mean      17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624     17.025624\n",
      "std      355.577741    403.040500    414.203986    696.187167    511.292732    563.492163    552.080294    605.202152    342.199733    467.907440    366.656325    259.470683    456.865423    388.310452    614.858325    487.820932    519.312056    738.871944    482.500835    296.348413    285.776934    692.354887    449.325128    414.200826    315.138256    515.491677    326.549859    720.550389    341.762458    295.832030    450.169131    630.474956    493.863197    581.309894    636.724788    517.106472    377.687587    445.250760    329.264408    457.984159    327.383124    417.886092    297.299311    446.439448    470.423602    480.281624    250.723864    225.128421    518.807529    464.715913\n",
      "\n",
      "================================================================================\n",
      "File: miRNA Expression Data\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/6miRNA_TPM.csv\n",
      "================================================================================\n",
      "Dimensions: 2201 rows x 51 columns\n",
      "Column names: ['sRNA', 'P102s', 'P105s', 'P111s', 'P113s', 'P117s', 'P118s', 'P123s', 'P127s', 'P128s']\n",
      "              ... (and 41 more columns)\n",
      "Data types:\n",
      "    float64: 50 columns\n",
      "    object: 1 columns\n",
      "\n",
      "First 5 rows:\n",
      "              sRNA        P102s        P105s        P111s        P113s        P117s        P118s        P123s        P127s        P128s\n",
      "0  hsa-let-7a-2-3p     0.000000     0.000000     0.000000     0.000000     0.000000     0.000000     0.000000     0.000000     0.161335\n",
      "1    hsa-let-7a-3p    30.341054    44.325241    42.349801    20.402847    27.398130    20.295304    32.510492    68.927690    25.490969\n",
      "2    hsa-let-7a-5p  3044.004813  3761.834535  5573.501486  5559.970235  5398.833298  9218.386371  4292.027489  9415.522432  5063.910136\n",
      "3    hsa-let-7b-3p     0.642819     0.675690     2.190507     1.943128     2.293797     1.295445     1.927500     3.759692     0.887344\n",
      "4    hsa-let-7b-5p   153.248035   347.304478   355.592291   298.173042   276.912259   369.979082   268.179437  1133.199071   165.852636\n",
      "\n",
      "No missing values found.\n",
      "\n",
      "Numeric columns summary:\n",
      "             P102s        P105s        P111s        P113s        P117s        P118s        P123s        P127s        P128s        P131s        P140s        P146s        P153s        P163s        P185s        P191s        P197s        P204s        P207s        P211s        P219s        P226s        P241s        P244s        P246s        P257s        P261s        P262s        P266s         P26s         P31s         P33s         P37s         P42s         P51s         P52s         P55s         P59s         P60s         P64s         P66s         P68s         P73s         P78s         P81s         P86s         P87s         P96s         P98s         P99s\n",
      "count  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000  2201.000000\n",
      "mean    454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937   454.338937\n",
      "std    5167.701141  5995.165579  4880.671522  5323.142202  4892.909508  5156.583113  5414.719362  4926.675995  5427.276178  4813.241468  5539.704911  5352.164770  5442.188191  5397.283751  5383.863969  4587.032674  5344.515415  5662.417153  4881.042343  5333.246440  5448.450550  4982.170529  4742.080807  5245.591508  4767.299862  5253.447453  4469.621014  4480.024060  4439.062992  5219.621893  4748.359239  4645.677617  5906.169780  4633.944809  5455.605137  5649.973352  4611.828416  5361.380144  4582.445819  5137.228156  4468.716562  5632.617386  5158.266182  4793.347187  5352.890140  5203.370200  6022.854651  5108.345166  5088.252347  4792.992519\n",
      "\n",
      "================================================================================\n",
      "File: CpG Methylation Data\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/1_PD_PromoterRegion_CpGs.csv\n",
      "================================================================================\n",
      "Dimensions: 17584 rows x 51 columns\n",
      "Column names: ['Gene_Symbol', 'P102d', 'P105d', 'P111d', 'P113d', 'P117d', 'P118d', 'P123d', 'P127d', 'P128d']\n",
      "              ... (and 41 more columns)\n",
      "Data types:\n",
      "    float64: 50 columns\n",
      "    object: 1 columns\n",
      "\n",
      "First 5 rows:\n",
      "  Gene_Symbol     P102d     P105d     P111d     P113d     P117d     P118d     P123d     P127d     P128d\n",
      "0        A1BG  0.582906  0.456747  0.535035  0.638074  0.615732  0.505413  0.554163  0.541218  0.586789\n",
      "1    A1BG-AS1  0.500771  0.376418  0.397345  0.521447  0.550180  0.339344  0.444468  0.453535  0.412788\n",
      "2     A2M-AS1  0.666854  0.713421  0.561839  0.573033  0.694567  0.603831  0.629768  0.699405  0.587279\n",
      "3       A2ML1  0.808487  0.788236  0.829482  0.878738  0.859099  0.831974  0.850209  0.821716  0.862571\n",
      "4   A2ML1-AS1  0.834131  0.876004  0.814006  0.841255  0.826644  0.819913  0.850567  0.908111  0.844114\n",
      "\n",
      "Missing values:\n",
      "    P26d: 17498 (99.5%)\n",
      "\n",
      "Numeric columns summary:\n",
      "              P102d         P105d         P111d         P113d         P117d         P118d         P123d         P127d         P128d         P131d         P140d         P146d         P153d         P163d         P185d         P191d         P197d         P204d         P207d         P211d         P219d         P226d         P241d         P244d         P246d         P257d         P261d         P262d         P266d       P26d          P31d          P33d          P37d          P42d          P51d          P52d          P55d          P59d          P60d          P64d          P66d          P68d          P73d          P78d          P81d          P86d          P87d          P96d          P98d          P99d\n",
      "count  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  86.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000  17584.000000\n",
      "mean       0.692308      0.677966      0.671963      0.692731      0.682784      0.678254      0.704739      0.683600      0.687860      0.694878      0.711281      0.671372      0.688742      0.677360      0.687840      0.690467      0.697571      0.690668      0.687472      0.689373      0.685854      0.683229      0.685846      0.689951      0.677754      0.670444      0.672706      0.687323      0.678036   0.635527      0.703103      0.685770      0.685921      0.684540      0.700385      0.685741      0.690129      0.676878      0.693817      0.692159      0.687807      0.693036      0.698260      0.695946      0.690609      0.680347      0.683851      0.695954      0.669736      0.683721\n",
      "std        0.196265      0.203752      0.204644      0.189351      0.190757      0.200477      0.194784      0.195528      0.201479      0.203754      0.194102      0.209691      0.199436      0.210318      0.199717      0.195771      0.195329      0.190720      0.207449      0.203778      0.197371      0.196872      0.204115      0.188439      0.197387      0.203305      0.205386      0.192935      0.202232   0.192234      0.193767      0.201512      0.194035      0.197675      0.192002      0.205907      0.208370      0.214389      0.197437      0.206265      0.196215      0.192559      0.197992      0.199927      0.195208      0.210970      0.198526      0.194767      0.203477      0.202934\n",
      "\n",
      "================================================================================\n",
      "File: Clinical Data\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/7Clinical_data50.csv\n",
      "================================================================================\n",
      "Dimensions: 50 rows x 21 columns\n",
      "Column names: ['NTUID', 'DNA_ID', 'mRNA_ID', 'sRNA_ID', 'age', 'BMI', 'BMI_cat', 'B12_status', 'B12_mol', 'B12supplem']\n",
      "              ... (and 11 more columns)\n",
      "Data types:\n",
      "    float64: 8 columns\n",
      "    object: 7 columns\n",
      "    int64: 6 columns\n",
      "\n",
      "First 5 rows:\n",
      "   NTUID DNA_ID mRNA_ID sRNA_ID   age        BMI     BMI_cat B12_status   B12_mol  B12supplem\n",
      "0    102  P102d   P102m   P102s  36.5  20.703735     healthy         NB  292.6170         1.0\n",
      "1    105  P105d   P105m   P105s  33.5  38.404033      obese2         NB  362.2104         1.0\n",
      "2    111  P111d   P111m   P111s  32.6  22.519433     healthy         NB  312.3954         1.0\n",
      "3    113  P113d   P113m   P113s  31.5  25.970116  overweight         NB  469.2204         1.0\n",
      "4    117  P117d   P117m   P117s  37.0  26.370238  overweight         LB  138.3012         2.0\n",
      "\n",
      "Missing values:\n",
      "    B12supplem: 2 (4.0%)\n",
      "    V1_se_EmploymentStatus: 4 (8.0%)\n",
      "    v1p_MultivitTab: 4 (8.0%)\n",
      "    v1p_FolicAcid: 2 (4.0%)\n",
      "    smoking: 5 (10.0%)\n",
      "\n",
      "Numeric columns summary:\n",
      "            NTUID        age        BMI     B12_mol  B12supplem     parity  V1_se_EmploymentStatus  V1_se_HouseholdIncome  v1p_MultivitTab  v1p_FolicAcid    smoking  v3n_Gender  batch_miRNA  batch_mRNA\n",
      "count   50.000000  50.000000  50.000000   50.000000   48.000000  50.000000               46.000000              50.000000        46.000000      48.000000  45.000000   50.000000    50.000000    50.00000\n",
      "mean   130.720000  30.654000  29.631478  244.770984    1.354167   1.560000                3.891304               7.540000         1.326087       1.145833   0.622222    1.460000     3.620000     1.56000\n",
      "std     72.925749   5.187316   7.732081  141.089535    0.483321   1.296148               14.114182               4.554478         0.473960       0.356674   0.911764    0.503457     1.576485     1.01338\n",
      "\n",
      "================================================================================\n",
      "File: Candidate Triplets\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_candidate_triplets.csv\n",
      "================================================================================\n",
      "Dimensions: 282 rows x 3 columns\n",
      "Column names: ['CpG', 'miRNA', 'mRNA']\n",
      "Data types:\n",
      "    object: 3 columns\n",
      "\n",
      "First 5 rows:\n",
      "           CpG            miRNA     mRNA\n",
      "0  CpG_PLEKHG4   hsa-miR-27a-5p  PLEKHG4\n",
      "1  CpG_PLEKHG4  hsa-miR-365a-5p  PLEKHG4\n",
      "2  CpG_PLEKHG4    hsa-miR-93-3p  PLEKHG4\n",
      "3    CpG_SCML2    hsa-miR-1269a    SCML2\n",
      "4    CpG_SCML2   hsa-miR-18b-5p    SCML2\n",
      "\n",
      "No missing values found.\n",
      "\n",
      "================================================================================\n",
      "File: DNA-mRNA Integration\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/9Integrated_DNA_mRNA.xlsx\n",
      "================================================================================\n",
      "Dimensions: 2762 rows x 38 columns\n",
      "Column names: ['Unnamed: 0', 'Gene.stable.ID', 'annot.tx_id', 'seqnames', 'start', 'end', 'width', 'strand', 'name', 'pvalue.x']\n",
      "              ... (and 28 more columns)\n",
      "Data types:\n",
      "    object: 18 columns\n",
      "    int64: 10 columns\n",
      "    float64: 10 columns\n",
      "\n",
      "First 5 rows:\n",
      "   Unnamed: 0   Gene.stable.ID        annot.tx_id seqnames      start        end  width strand name      pvalue.x\n",
      "0           1  ENSG00000003436  ENST00000233156.9     chr2  187540002  187541000    999      *    *  1.314200e-03\n",
      "1           2  ENSG00000003436  ENST00000233156.9     chr2  187556002  187557000    999      *    *  1.760000e-09\n",
      "2           3  ENSG00000003436  ENST00000437725.5     chr2  187556002  187557000    999      *    *  1.760000e-09\n",
      "3           4  ENSG00000003436  ENST00000421427.5     chr2  187540002  187541000    999      *    *  1.314200e-03\n",
      "4           5  ENSG00000003436  ENST00000417013.5     chr2  187556002  187557000    999      *    *  1.760000e-09\n",
      "\n",
      "Missing values:\n",
      "    annot.gene_id: 10 (0.4%)\n",
      "    annot.symbol: 10 (0.4%)\n",
      "    HGNC.symbol: 4 (0.1%)\n",
      "    SYMBOL: 5 (0.2%)\n",
      "\n",
      "Numeric columns summary:\n",
      "        Unnamed: 0         start           end   width     pvalue.x       qvalue    meth.diff   annot.start     annot.end   annot.width  annot.gene_id      baseMean  log2FoldChange        lfcSE         stat     pvalue.y         padj    Gene_start      Gene_end       Strand\n",
      "count  2762.000000  2.762000e+03  2.762000e+03  2762.0  2762.000000  2762.000000  2762.000000  2.762000e+03  2.762000e+03   2762.000000   2.752000e+03   2762.000000     2762.000000  2762.000000  2762.000000  2762.000000  2762.000000  2.762000e+03  2.762000e+03  2762.000000\n",
      "mean   1381.500000  6.816971e+07  6.817071e+07   999.0     0.002807     0.006232     0.634957  6.816136e+07  6.818183e+07  20464.811369   3.726437e+05   1783.199635        2.311394     0.707373     1.460588     0.000133     0.019040  6.812116e+07  6.826673e+07     0.230992\n",
      "std     797.465046  4.868208e+07  4.868208e+07     0.0     0.005868     0.012252     2.529406  4.867869e+07  4.868524e+07  50220.894193   5.900282e+06  11641.300658        7.531798     1.184378     3.890644     0.000122     0.014123  4.865159e+07  4.870243e+07     0.973132\n",
      "\n",
      "================================================================================\n",
      "File: miRNA-mRNA Integration\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/10Integrated_miRNA_mRNA.xlsx\n",
      "================================================================================\n",
      "Dimensions: 5981 rows x 11 columns\n",
      "Column names: ['Target Gene', 'mirna', 'log2FoldChange_mirna', 'pvalue_mirna', 'miRTarBase ID', 'mrna', 'log2FoldChange_mrna', 'pvalue_mrna', 'padj_mrna', 'Gene_type']\n",
      "              ... (and 1 more columns)\n",
      "Data types:\n",
      "    object: 6 columns\n",
      "    float64: 5 columns\n",
      "\n",
      "First 5 rows:\n",
      "  Target Gene         mirna  log2FoldChange_mirna  pvalue_mirna miRTarBase ID             mrna  log2FoldChange_mrna   pvalue_mrna     padj_mrna       Gene_type\n",
      "0     C4orf50  hsa-miR-1299             -1.857076       0.00779           NaN  ENSG00000181215            30.000000  7.470000e-14  2.790000e-10  protein_coding\n",
      "1        SGCZ  hsa-miR-1299             -1.857076       0.00779           NaN  ENSG00000185053            24.196848  1.570000e-09  1.830000e-06  protein_coding\n",
      "2      SHISA9  hsa-miR-1299             -1.857076       0.00779           NaN  ENSG00000237515            20.752643  2.270000e-07  1.376650e-04  protein_coding\n",
      "3      SHISA9  hsa-miR-1299             -1.857076       0.00779           NaN  ENSG00000237515            20.752643  2.270000e-07  1.376650e-04  protein_coding\n",
      "4      SHISA9  hsa-miR-1299             -1.857076       0.00779           NaN  ENSG00000237515            20.752643  2.270000e-07  1.376650e-04  protein_coding\n",
      "\n",
      "Missing values:\n",
      "    miRTarBase ID: 5753 (96.2%)\n",
      "\n",
      "Numeric columns summary:\n",
      "       log2FoldChange_mirna  pvalue_mirna  log2FoldChange_mrna  pvalue_mrna    padj_mrna\n",
      "count           5981.000000   5981.000000          5981.000000  5981.000000  5981.000000\n",
      "mean               2.194705      0.022968             2.551128     0.000166     0.022630\n",
      "std                6.688761      0.013631             7.639380     0.000134     0.015238\n",
      "\n",
      "================================================================================\n",
      "File: 3DNA Annotation\n",
      "Path: /Users/heweilin/Desktop/P056_Code_2/Raw_Data/3DNA_all.csv\n",
      "================================================================================\n",
      "Dimensions: 1046209 rows x 24 columns\n",
      "Column names: ['Unnamed: 0', 'annot.tx_id', 'seqnames', 'start', 'end', 'width', 'strand', 'name', 'pvalue', 'qvalue']\n",
      "              ... (and 14 more columns)\n",
      "Data types:\n",
      "    object: 13 columns\n",
      "    int64: 7 columns\n",
      "    float64: 4 columns\n",
      "\n",
      "First 5 rows:\n",
      "   Unnamed: 0         annot.tx_id seqnames      start        end  width strand name    pvalue    qvalue\n",
      "0           1  ENST00000000233.10     chr7  127590002  127591000    999      *    *  0.000005  0.000023\n",
      "1           2  ENST00000000233.10     chr7  127590002  127591000    999      *    *  0.000005  0.000023\n",
      "2           3  ENST00000000233.10     chr7  127590002  127591000    999      *    *  0.000005  0.000023\n",
      "3           4  ENST00000000233.10     chr7  127590002  127591000    999      *    *  0.000005  0.000023\n",
      "4           5  ENST00000000442.11    chr11   64314002   64315000    999      *    *  0.042895  0.079577\n",
      "\n",
      "Missing values:\n",
      "    annot.tx_id: 125089 (12.0%)\n",
      "    annot.gene_id: 156506 (15.0%)\n",
      "    annot.symbol: 156560 (15.0%)\n",
      "    Gene.stable.ID: 125089 (12.0%)\n",
      "    Gene.stable.ID.version: 125089 (12.0%)\n",
      "    Transcript.stable.ID: 125089 (12.0%)\n",
      "    HGNC.symbol: 174790 (16.7%)\n",
      "\n",
      "Numeric columns summary:\n",
      "         Unnamed: 0         start           end      width        pvalue        qvalue     meth.diff   annot.start     annot.end   annot.width  annot.gene_id\n",
      "count  1.046209e+06  1.046209e+06  1.046209e+06  1046209.0  1.046209e+06  1.046209e+06  1.046209e+06  1.046209e+06  1.046209e+06  1.046209e+06   8.897030e+05\n",
      "mean   5.231050e+05  6.741120e+07  6.741220e+07      999.0  2.040580e-01  2.432344e-01  3.603416e-01  6.738966e+07  6.743391e+07  4.425690e+04   6.957798e+06\n",
      "std    3.020147e+05  5.695768e+07  5.695768e+07        0.0  2.854555e-01  3.078657e-01  1.751465e+00  5.695326e+07  5.696293e+07  2.299548e+05   2.612526e+07\n",
      "\n",
      "================================================================================\n",
      "SAMPLE CONSISTENCY CHECK\n",
      "================================================================================\n",
      "mRNA samples: 50 (first 5: ['P26m', 'P31m', 'P33m', 'P37m', 'P42m'])\n",
      "miRNA samples: 50 (first 5: ['P102s', 'P105s', 'P111s', 'P113s', 'P117s'])\n",
      "CpG samples: 50 (first 5: ['P102d', 'P105d', 'P111d', 'P113d', 'P117d'])\n",
      "\n",
      "Common samples across all matrices: 0\n",
      "mRNA unique samples: 50\n",
      "miRNA unique samples: 50\n",
      "CpG unique samples: 50\n",
      "\n",
      "================================================================================\n",
      "TRIPLET COVERAGE CHECK\n",
      "================================================================================\n",
      "Total triplets: 282\n",
      "\n",
      "mRNA Coverage:\n",
      "    Available features: 58735\n",
      "    Required by triplets: 41\n",
      "    Covered: 0 (0.0%)\n",
      "    Missing: 41\n",
      "    Missing features (first 10): ['PLEKHG4', 'SCML2', 'LEP', 'LHX3', 'FBXO17', 'IFITM10', 'RAI14', 'TAB3', 'CGB8', 'FAM167A']\n",
      "\n",
      "miRNA Coverage:\n",
      "    Available features: 2201\n",
      "    Required by triplets: 41\n",
      "    Covered: 41 (100.0%)\n",
      "    Missing: 0\n",
      "\n",
      "CpG Coverage:\n",
      "    Available features: 17584\n",
      "    Required by triplets: 41\n",
      "    Covered: 0 (0.0%)\n",
      "    Missing: 41\n",
      "    Missing features (first 10): ['CpG_TTLL10', 'CpG_GEM', 'CpG_CYP2C19', 'CpG_LEP', 'CpG_HYAL2', 'CpG_ZBTB16', 'CpG_GYPC', 'CpG_DMTN', 'CpG_PARP3', 'CpG_CGB5']\n",
      "\n",
      "================================================================================\n",
      "PREVIEW COMPLETE\n",
      "================================================================================\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Set file paths\n",
    "base_dir = \"/Users/heweilin/Desktop/P056_Code_2\"\n",
    "raw_data_dir = os.path.join(base_dir, \"Raw_Data\")\n",
    "processed_data_dir = os.path.join(base_dir, \"Processed_Data\")\n",
    "\n",
    "# Define all files to check\n",
    "files_to_check = {\n",
    "    \"mRNA Expression Data\": os.path.join(raw_data_dir, \"5mRNA_TPM.csv\"),\n",
    "    \"miRNA Expression Data\": os.path.join(raw_data_dir, \"6miRNA_TPM.csv\"),\n",
    "    \"CpG Methylation Data\": os.path.join(processed_data_dir, \"1_PD_PromoterRegion_CpGs.csv\"),\n",
    "    \"Clinical Data\": os.path.join(raw_data_dir, \"7Clinical_data50.csv\"),\n",
    "    \"Candidate Triplets\": os.path.join(processed_data_dir, \"3_candidate_triplets.csv\"),\n",
    "    \"DNA-mRNA Integration\": os.path.join(raw_data_dir, \"9Integrated_DNA_mRNA.xlsx\"),\n",
    "    \"miRNA-mRNA Integration\": os.path.join(raw_data_dir, \"10Integrated_miRNA_mRNA.xlsx\"),\n",
    "    \"3DNA Annotation\": os.path.join(raw_data_dir, \"3DNA_all.csv\")\n",
    "}\n",
    "\n",
    "def preview_file(file_path, file_name, max_cols=10):\n",
    "    \"\"\"Preview basic information and first 5 rows of a single file\"\"\"\n",
    "    print(f\"\\n{'='*80}\")\n",
    "    print(f\"File: {file_name}\")\n",
    "    print(f\"Path: {file_path}\")\n",
    "    print(f\"{'='*80}\")\n",
    "    \n",
    "    if not os.path.exists(file_path):\n",
    "        print(\"ERROR: File does not exist!\")\n",
    "        return\n",
    "    \n",
    "    try:\n",
    "        # Choose reading method based on file extension\n",
    "        if file_path.endswith('.xlsx'):\n",
    "            data = pd.read_excel(file_path)\n",
    "        elif file_path.endswith('.csv'):\n",
    "            data = pd.read_csv(file_path)\n",
    "        else:\n",
    "            print(\"ERROR: Unsupported file format!\")\n",
    "            return\n",
    "        \n",
    "        # Display basic information\n",
    "        print(f\"Dimensions: {data.shape[0]} rows x {data.shape[1]} columns\")\n",
    "        print(f\"Column names: {list(data.columns[:max_cols])}\")\n",
    "        if data.shape[1] > max_cols:\n",
    "            print(f\"              ... (and {data.shape[1] - max_cols} more columns)\")\n",
    "        \n",
    "        # Display data types\n",
    "        print(f\"Data types:\")\n",
    "        dtype_summary = data.dtypes.value_counts()\n",
    "        for dtype, count in dtype_summary.items():\n",
    "            print(f\"    {dtype}: {count} columns\")\n",
    "        \n",
    "        # Display first 5 rows (limit columns for display)\n",
    "        print(f\"\\nFirst 5 rows:\")\n",
    "        display_data = data.iloc[:5, :max_cols] if data.shape[1] > max_cols else data.head()\n",
    "        \n",
    "        # Format display\n",
    "        pd.set_option('display.max_columns', max_cols)\n",
    "        pd.set_option('display.width', 120)\n",
    "        pd.set_option('display.max_colwidth', 15)\n",
    "        \n",
    "        print(display_data.to_string())\n",
    "        \n",
    "        # Check missing values\n",
    "        missing_summary = data.isnull().sum()\n",
    "        missing_cols = missing_summary[missing_summary > 0]\n",
    "        if len(missing_cols) > 0:\n",
    "            print(f\"\\nMissing values:\")\n",
    "            for col, missing_count in missing_cols.head(10).items():\n",
    "                missing_pct = (missing_count / len(data)) * 100\n",
    "                print(f\"    {col}: {missing_count} ({missing_pct:.1f}%)\")\n",
    "            if len(missing_cols) > 10:\n",
    "                print(f\"    ... (and {len(missing_cols) - 10} more columns with missing values)\")\n",
    "        else:\n",
    "            print(f\"\\nNo missing values found.\")\n",
    "        \n",
    "        # Additional statistics for numeric columns\n",
    "        numeric_cols = data.select_dtypes(include=[np.number]).columns\n",
    "        if len(numeric_cols) > 0:\n",
    "            print(f\"\\nNumeric columns summary:\")\n",
    "            numeric_summary = data[numeric_cols].describe().iloc[:3]  # Only count, mean, std\n",
    "            print(numeric_summary.to_string())\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"ERROR reading file: {str(e)}\")\n",
    "\n",
    "def check_sample_consistency():\n",
    "    \"\"\"Check sample ID consistency across expression matrices\"\"\"\n",
    "    print(f\"\\n{'='*80}\")\n",
    "    print(\"SAMPLE CONSISTENCY CHECK\")\n",
    "    print(f\"{'='*80}\")\n",
    "    \n",
    "    sample_info = {}\n",
    "    \n",
    "    # Check expression matrices\n",
    "    expr_files = {\n",
    "        \"mRNA\": os.path.join(raw_data_dir, \"5mRNA_TPM.csv\"),\n",
    "        \"miRNA\": os.path.join(raw_data_dir, \"6miRNA_TPM.csv\"),\n",
    "        \"CpG\": os.path.join(processed_data_dir, \"1_PD_PromoterRegion_CpGs.csv\")\n",
    "    }\n",
    "    \n",
    "    for data_type, file_path in expr_files.items():\n",
    "        if os.path.exists(file_path):\n",
    "            try:\n",
    "                data = pd.read_csv(file_path, index_col=0)\n",
    "                sample_info[data_type] = set(data.columns)\n",
    "                print(f\"{data_type} samples: {len(data.columns)} (first 5: {list(data.columns[:5])})\")\n",
    "            except Exception as e:\n",
    "                print(f\"Error reading {data_type}: {e}\")\n",
    "    \n",
    "    # Find common samples\n",
    "    if len(sample_info) > 1:\n",
    "        common_samples = set.intersection(*sample_info.values())\n",
    "        print(f\"\\nCommon samples across all matrices: {len(common_samples)}\")\n",
    "        \n",
    "        # Show sample differences\n",
    "        for data_type, samples in sample_info.items():\n",
    "            unique_samples = samples - common_samples\n",
    "            if unique_samples:\n",
    "                print(f\"{data_type} unique samples: {len(unique_samples)}\")\n",
    "\n",
    "def check_triplet_coverage():\n",
    "    \"\"\"Check if triplet features are present in expression data\"\"\"\n",
    "    print(f\"\\n{'='*80}\")\n",
    "    print(\"TRIPLET COVERAGE CHECK\")\n",
    "    print(f\"{'='*80}\")\n",
    "    \n",
    "    # Load triplets\n",
    "    triplets_path = os.path.join(processed_data_dir, \"3_candidate_triplets.csv\")\n",
    "    if not os.path.exists(triplets_path):\n",
    "        print(\"Triplets file not found!\")\n",
    "        return\n",
    "    \n",
    "    triplets = pd.read_csv(triplets_path)\n",
    "    print(f\"Total triplets: {len(triplets)}\")\n",
    "    \n",
    "    # Check coverage for each data type\n",
    "    coverage_info = {}\n",
    "    \n",
    "    # Check mRNA coverage\n",
    "    mrna_path = os.path.join(raw_data_dir, \"5mRNA_TPM.csv\")\n",
    "    if os.path.exists(mrna_path):\n",
    "        mrna_data = pd.read_csv(mrna_path, index_col=0)\n",
    "        mrna_genes = set(mrna_data.index)\n",
    "        triplet_mrnas = set(triplets['mRNA'])\n",
    "        coverage_info['mRNA'] = {\n",
    "            'available': len(mrna_genes),\n",
    "            'required': len(triplet_mrnas),\n",
    "            'covered': len(triplet_mrnas & mrna_genes),\n",
    "            'missing': triplet_mrnas - mrna_genes\n",
    "        }\n",
    "    \n",
    "    # Check miRNA coverage\n",
    "    mirna_path = os.path.join(raw_data_dir, \"6miRNA_TPM.csv\")\n",
    "    if os.path.exists(mirna_path):\n",
    "        mirna_data = pd.read_csv(mirna_path, index_col=0)\n",
    "        mirna_features = set(mirna_data.index)\n",
    "        triplet_mirnas = set(triplets['miRNA'])\n",
    "        coverage_info['miRNA'] = {\n",
    "            'available': len(mirna_features),\n",
    "            'required': len(triplet_mirnas),\n",
    "            'covered': len(triplet_mirnas & mirna_features),\n",
    "            'missing': triplet_mirnas - mirna_features\n",
    "        }\n",
    "    \n",
    "    # Check CpG coverage\n",
    "    cpg_path = os.path.join(processed_data_dir, \"1_PD_PromoterRegion_CpGs.csv\")\n",
    "    if os.path.exists(cpg_path):\n",
    "        cpg_data = pd.read_csv(cpg_path, index_col=0)\n",
    "        cpg_features = set(cpg_data.index)\n",
    "        triplet_cpgs = set(triplets['CpG'])\n",
    "        coverage_info['CpG'] = {\n",
    "            'available': len(cpg_features),\n",
    "            'required': len(triplet_cpgs),\n",
    "            'covered': len(triplet_cpgs & cpg_features),\n",
    "            'missing': triplet_cpgs - cpg_features\n",
    "        }\n",
    "    \n",
    "    # Display coverage results\n",
    "    for data_type, info in coverage_info.items():\n",
    "        coverage_pct = (info['covered'] / info['required']) * 100 if info['required'] > 0 else 0\n",
    "        print(f\"\\n{data_type} Coverage:\")\n",
    "        print(f\"    Available features: {info['available']}\")\n",
    "        print(f\"    Required by triplets: {info['required']}\")\n",
    "        print(f\"    Covered: {info['covered']} ({coverage_pct:.1f}%)\")\n",
    "        print(f\"    Missing: {len(info['missing'])}\")\n",
    "        \n",
    "        if len(info['missing']) > 0 and len(info['missing']) <= 10:\n",
    "            print(f\"    Missing features: {list(info['missing'])}\")\n",
    "        elif len(info['missing']) > 10:\n",
    "            print(f\"    Missing features (first 10): {list(info['missing'])[:10]}\")\n",
    "\n",
    "def main():\n",
    "    \"\"\"Main function to preview all data files\"\"\"\n",
    "    print(\"DATA FILES PREVIEW REPORT\")\n",
    "    print(\"=\" * 80)\n",
    "    \n",
    "    # Preview each file\n",
    "    for file_name, file_path in files_to_check.items():\n",
    "        preview_file(file_path, file_name)\n",
    "    \n",
    "    # Check sample consistency\n",
    "    check_sample_consistency()\n",
    "    \n",
    "    # Check triplet coverage\n",
    "    check_triplet_coverage()\n",
    "    \n",
    "    print(f\"\\n{'='*80}\")\n",
    "    print(\"PREVIEW COMPLETE\")\n",
    "    print(f\"{'='*80}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "026d1a3e-f50f-4e67-a5c0-4bddf3ce56ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "EXPRESSION DATA STANDARDIZATION AND COVARIATE ADJUSTMENT\n",
      "================================================================================\n",
      "================================================================================\n",
      "DATA LOADING AND PREPROCESSING\n",
      "================================================================================\n",
      "1. Loading mRNA expression data...\n",
      "   mRNA data: (58735, 50)\n",
      "   Sample examples: ['P26m', 'P31m', 'P33m', 'P37m', 'P42m']\n",
      "2. Loading miRNA expression data...\n",
      "   miRNA data: (2201, 50)\n",
      "   Sample examples: ['P102s', 'P105s', 'P111s', 'P113s', 'P117s']\n",
      "3. Loading CpG methylation data...\n",
      "   CpG data: (17584, 50)\n",
      "   Sample examples: ['P102d', 'P105d', 'P111d', 'P113d', 'P117d']\n",
      "4. Loading clinical data...\n",
      "   Clinical data: (50, 21)\n",
      "   Key columns: ['NTUID', 'DNA_ID', 'mRNA_ID', 'sRNA_ID', 'age', 'BMI', 'BMI_cat', 'B12_status', 'B12_mol', 'B12supplem']\n",
      "5. Loading candidate triplets...\n",
      "   Number of triplets: 282\n",
      "6. Loading DNA-mRNA integration data...\n",
      "   Integration data: (2762, 38)\n",
      "\n",
      "==================================================\n",
      "STANDARDIZING SAMPLE IDs\n",
      "==================================================\n",
      "Created 150 sample ID mappings\n",
      "\n",
      "Creating gene ID mapping...\n",
      "Created 111 gene ID mappings\n",
      "\n",
      "Harmonizing sample names...\n",
      "Processing mRNA data...\n",
      "   mRNA: (58735, 50) -> (58735, 50)\n",
      "   Retained samples: ['P26', 'P31', 'P33', 'P37', 'P42']...\n",
      "Processing miRNA data...\n",
      "   miRNA: (2201, 50) -> (2201, 50)\n",
      "   Retained samples: ['P102', 'P105', 'P111', 'P113', 'P117']...\n",
      "Processing CpG data...\n",
      "   CpG: (17584, 50) -> (17584, 49)\n",
      "   Retained samples: ['P102', 'P105', 'P111', 'P113', 'P117']...\n",
      "\n",
      "Common sample count: 49\n",
      "\n",
      "Creating CpG-gene mapping...\n",
      "Mapped 41 CpG sites\n",
      "\n",
      "==================================================\n",
      "FILTERING TRIPLET-RELEVANT FEATURES\n",
      "==================================================\n",
      "1. Filtering mRNA features...\n",
      "   mRNA: 41 required -> 41 available\n",
      "2. Filtering miRNA features...\n",
      "   miRNA: 41 required -> 41 available\n",
      "3. Filtering CpG features...\n",
      "   CpG: 41 required -> 41 available\n",
      "\n",
      "Preparing covariate data...\n",
      "   Covariate matrix: (49, 6)\n",
      "   Covariate columns: ['age', 'BMI', 'parity', 'batch_miRNA', 'batch_mRNA', 'B12_NB']\n",
      "\n",
      "Processing mRNA data...\n",
      "\n",
      "Standardizing data using log2_zscore method...\n",
      "Adjusting 41 features for covariates...\n",
      "\n",
      "Processing miRNA data...\n",
      "\n",
      "Standardizing data using log2_zscore method...\n",
      "Adjusting 41 features for covariates...\n",
      "\n",
      "Processing CpG data...\n",
      "\n",
      "Standardizing data using zscore_only method...\n",
      "\n",
      "==================================================\n",
      "SAVING PROCESSED DATA\n",
      "==================================================\n",
      "Saved mRNA: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_mRNA_expr_adjusted.csv ((41, 49))\n",
      "Saved miRNA: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_miRNA_expr_adjusted.csv ((41, 49))\n",
      "Saved CpG: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_CpG_expr_adjusted.csv ((41, 49))\n",
      "Processing report: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_processing_report.txt\n",
      "\n",
      "================================================================================\n",
      "PROCESSING COMPLETE!\n",
      "Output files:\n",
      "  mRNA: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_mRNA_expr_adjusted.csv\n",
      "  miRNA: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_miRNA_expr_adjusted.csv\n",
      "  CpG: /Users/heweilin/Desktop/P056_Code_2/Processed_Data/3_CpG_expr_adjusted.csv\n",
      "================================================================================\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy import stats\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import LinearRegression\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# Set file paths\n",
    "base_dir = \"/Users/heweilin/Desktop/P056_Code_2\"\n",
    "raw_data_dir = os.path.join(base_dir, \"Raw_Data\")\n",
    "processed_data_dir = os.path.join(base_dir, \"Processed_Data\")\n",
    "\n",
    "# Input file paths\n",
    "mrna_path = os.path.join(raw_data_dir, \"5mRNA_TPM.csv\")\n",
    "mirna_path = os.path.join(raw_data_dir, \"6miRNA_TPM.csv\")\n",
    "cpg_path = os.path.join(processed_data_dir, \"1_PD_PromoterRegion_CpGs.csv\")\n",
    "clinical_path = os.path.join(raw_data_dir, \"7Clinical_data50.csv\")\n",
    "triplets_path = os.path.join(processed_data_dir, \"3_candidate_triplets.csv\")\n",
    "dna_mrna_path = os.path.join(raw_data_dir, \"9Integrated_DNA_mRNA.xlsx\")\n",
    "\n",
    "# Ensure output directory exists\n",
    "os.makedirs(processed_data_dir, exist_ok=True)\n",
    "\n",
    "def load_and_prepare_data():\n",
    "    \"\"\"Load and preprocess all data files\"\"\"\n",
    "    print(\"=\" * 80)\n",
    "    print(\"DATA LOADING AND PREPROCESSING\")\n",
    "    print(\"=\" * 80)\n",
    "    \n",
    "    # 1. Load mRNA expression data\n",
    "    print(\"1. Loading mRNA expression data...\")\n",
    "    mrna_data = pd.read_csv(mrna_path)\n",
    "    # Handle first column as gene ID\n",
    "    if 'Unnamed: 0' in mrna_data.columns:\n",
    "        mrna_data = mrna_data.set_index('Unnamed: 0')\n",
    "    else:\n",
    "        mrna_data = mrna_data.set_index(mrna_data.columns[0])\n",
    "    print(f\"   mRNA data: {mrna_data.shape}\")\n",
    "    print(f\"   Sample examples: {list(mrna_data.columns[:5])}\")\n",
    "    \n",
    "    # 2. Load miRNA expression data\n",
    "    print(\"2. Loading miRNA expression data...\")\n",
    "    mirna_data = pd.read_csv(mirna_path)\n",
    "    if 'sRNA' in mirna_data.columns:\n",
    "        mirna_data = mirna_data.set_index('sRNA')\n",
    "    else:\n",
    "        mirna_data = mirna_data.set_index(mirna_data.columns[0])\n",
    "    print(f\"   miRNA data: {mirna_data.shape}\")\n",
    "    print(f\"   Sample examples: {list(mirna_data.columns[:5])}\")\n",
    "    \n",
    "    # 3. Load CpG methylation data\n",
    "    print(\"3. Loading CpG methylation data...\")\n",
    "    cpg_data = pd.read_csv(cpg_path)\n",
    "    if 'Gene_Symbol' in cpg_data.columns:\n",
    "        cpg_data = cpg_data.set_index('Gene_Symbol')\n",
    "    else:\n",
    "        cpg_data = cpg_data.set_index(cpg_data.columns[0])\n",
    "    print(f\"   CpG data: {cpg_data.shape}\")\n",
    "    print(f\"   Sample examples: {list(cpg_data.columns[:5])}\")\n",
    "    \n",
    "    # 4. Load clinical data\n",
    "    print(\"4. Loading clinical data...\")\n",
    "    clinical_data = pd.read_csv(clinical_path)\n",
    "    print(f\"   Clinical data: {clinical_data.shape}\")\n",
    "    print(f\"   Key columns: {list(clinical_data.columns[:10])}\")\n",
    "    \n",
    "    # 5. Load candidate triplets\n",
    "    print(\"5. Loading candidate triplets...\")\n",
    "    triplets = pd.read_csv(triplets_path)\n",
    "    print(f\"   Number of triplets: {triplets.shape[0]}\")\n",
    "    \n",
    "    # 6. Load DNA-mRNA integration data (for gene ID conversion)\n",
    "    print(\"6. Loading DNA-mRNA integration data...\")\n",
    "    dna_mrna_integration = pd.read_excel(dna_mrna_path)\n",
    "    print(f\"   Integration data: {dna_mrna_integration.shape}\")\n",
    "    \n",
    "    return mrna_data, mirna_data, cpg_data, clinical_data, triplets, dna_mrna_integration\n",
    "\n",
    "def standardize_sample_ids():\n",
    "    \"\"\"Standardize sample ID mapping\"\"\"\n",
    "    print(\"\\n\" + \"=\" * 50)\n",
    "    print(\"STANDARDIZING SAMPLE IDs\")\n",
    "    print(\"=\" * 50)\n",
    "    \n",
    "    # Create sample ID mapping dictionary\n",
    "    # Based on ID correspondence in clinical data\n",
    "    clinical_data = pd.read_csv(clinical_path)\n",
    "    \n",
    "    sample_mapping = {}\n",
    "    for _, row in clinical_data.iterrows():\n",
    "        base_id = str(row['NTUID'])\n",
    "        sample_mapping[f\"P{base_id}m\"] = f\"P{base_id}\"  # mRNA\n",
    "        sample_mapping[f\"P{base_id}s\"] = f\"P{base_id}\"  # miRNA\n",
    "        sample_mapping[f\"P{base_id}d\"] = f\"P{base_id}\"  # CpG\n",
    "    \n",
    "    print(f\"Created {len(sample_mapping)} sample ID mappings\")\n",
    "    return sample_mapping\n",
    "\n",
    "def create_gene_id_mapping(dna_mrna_integration):\n",
    "    \"\"\"Create ENSEMBL ID to gene symbol mapping\"\"\"\n",
    "    print(\"\\nCreating gene ID mapping...\")\n",
    "    \n",
    "    # Extract gene ID mapping from integration data\n",
    "    gene_mapping = {}\n",
    "    \n",
    "    if 'Gene.stable.ID' in dna_mrna_integration.columns and 'annot.symbol' in dna_mrna_integration.columns:\n",
    "        mapping_df = dna_mrna_integration[['Gene.stable.ID', 'annot.symbol']].dropna()\n",
    "        gene_mapping = dict(zip(mapping_df['Gene.stable.ID'], mapping_df['annot.symbol']))\n",
    "    elif 'Gene.stable.ID' in dna_mrna_integration.columns and 'SYMBOL' in dna_mrna_integration.columns:\n",
    "        mapping_df = dna_mrna_integration[['Gene.stable.ID', 'SYMBOL']].dropna()\n",
    "        gene_mapping = dict(zip(mapping_df['Gene.stable.ID'], mapping_df['SYMBOL']))\n",
    "    \n",
    "    print(f\"Created {len(gene_mapping)} gene ID mappings\")\n",
    "    return gene_mapping\n",
    "\n",
    "def harmonize_sample_names(data_dict, sample_mapping):\n",
    "    \"\"\"Harmonize sample names across all datasets\"\"\"\n",
    "    print(\"\\nHarmonizing sample names...\")\n",
    "    \n",
    "    harmonized_data = {}\n",
    "    \n",
    "    for data_type, data in data_dict.items():\n",
    "        print(f\"Processing {data_type} data...\")\n",
    "        \n",
    "        # Rename columns\n",
    "        new_columns = []\n",
    "        for col in data.columns:\n",
    "            if col in sample_mapping:\n",
    "                new_columns.append(sample_mapping[col])\n",
    "            else:\n",
    "                new_columns.append(col)\n",
    "        \n",
    "        data_renamed = data.copy()\n",
    "        data_renamed.columns = new_columns\n",
    "        \n",
    "        # Remove samples with excessive missing values (e.g., P26)\n",
    "        missing_threshold = 0.8  # Remove if >80% values are missing\n",
    "        samples_to_keep = []\n",
    "        \n",
    "        for col in data_renamed.columns:\n",
    "            missing_rate = data_renamed[col].isna().sum() / len(data_renamed)\n",
    "            if missing_rate < missing_threshold:\n",
    "                samples_to_keep.append(col)\n",
    "        \n",
    "        data_clean = data_renamed[samples_to_keep]\n",
    "        harmonized_data[data_type] = data_clean\n",
    "        \n",
    "        print(f\"   {data_type}: {data.shape} -> {data_clean.shape}\")\n",
    "        print(f\"   Retained samples: {samples_to_keep[:5]}...\")\n",
    "    \n",
    "    return harmonized_data\n",
    "\n",
    "def create_cpg_gene_mapping(cpg_data, triplets):\n",
    "    \"\"\"Create gene to CpG mapping for CpG data\"\"\"\n",
    "    print(\"\\nCreating CpG-gene mapping...\")\n",
    "    \n",
    "    # Extract CpG-gene relationships from triplets\n",
    "    cpg_gene_mapping = {}\n",
    "    \n",
    "    for _, row in triplets.iterrows():\n",
    "        cpg_id = row['CpG']\n",
    "        gene_symbol = row['mRNA']\n",
    "        \n",
    "        # Try different CpG ID formats\n",
    "        possible_cpg_names = [\n",
    "            cpg_id,\n",
    "            cpg_id.replace('CpG_', ''),\n",
    "            gene_symbol,\n",
    "            f\"CpG_{gene_symbol}\"\n",
    "        ]\n",
    "        \n",
    "        found_cpg = None\n",
    "        for possible_name in possible_cpg_names:\n",
    "            if possible_name in cpg_data.index:\n",
    "                found_cpg = possible_name\n",
    "                break\n",
    "        \n",
    "        if found_cpg:\n",
    "            cpg_gene_mapping[cpg_id] = found_cpg\n",
    "        else:\n",
    "            # If no exact match, try partial matching\n",
    "            for cpg_index in cpg_data.index:\n",
    "                if gene_symbol.lower() in str(cpg_index).lower():\n",
    "                    cpg_gene_mapping[cpg_id] = cpg_index\n",
    "                    break\n",
    "    \n",
    "    print(f\"Mapped {len(cpg_gene_mapping)} CpG sites\")\n",
    "    return cpg_gene_mapping\n",
    "\n",
    "def filter_triplet_features(harmonized_data, triplets, gene_mapping, cpg_gene_mapping):\n",
    "    \"\"\"Filter features relevant to triplets\"\"\"\n",
    "    print(\"\\n\" + \"=\" * 50)\n",
    "    print(\"FILTERING TRIPLET-RELEVANT FEATURES\")\n",
    "    print(\"=\" * 50)\n",
    "    \n",
    "    filtered_data = {}\n",
    "    \n",
    "    # 1. Filter mRNA data\n",
    "    print(\"1. Filtering mRNA features...\")\n",
    "    mrna_data = harmonized_data['mRNA']\n",
    "    \n",
    "    # Get required genes\n",
    "    required_genes = set(triplets['mRNA'].unique())\n",
    "    \n",
    "    # Try direct gene symbol matching\n",
    "    available_genes_direct = set(gene_mapping.values()) & required_genes\n",
    "    \n",
    "    # Create reverse mapping (gene symbol -> ENSEMBL ID)\n",
    "    reverse_gene_mapping = {v: k for k, v in gene_mapping.items()}\n",
    "    \n",
    "    # Filter available mRNA features\n",
    "    mrna_features_to_keep = []\n",
    "    gene_symbol_mapping = {}\n",
    "    \n",
    "    for gene in required_genes:\n",
    "        if gene in reverse_gene_mapping:\n",
    "            ensembl_id = reverse_gene_mapping[gene]\n",
    "            if ensembl_id in mrna_data.index:\n",
    "                mrna_features_to_keep.append(ensembl_id)\n",
    "                gene_symbol_mapping[ensembl_id] = gene\n",
    "    \n",
    "    filtered_mrna = mrna_data.loc[mrna_features_to_keep]\n",
    "    # Rename index to gene symbols\n",
    "    filtered_mrna.index = [gene_symbol_mapping[idx] for idx in filtered_mrna.index]\n",
    "    \n",
    "    filtered_data['mRNA'] = filtered_mrna\n",
    "    print(f\"   mRNA: {len(required_genes)} required -> {len(filtered_mrna)} available\")\n",
    "    \n",
    "    # 2. Filter miRNA data\n",
    "    print(\"2. Filtering miRNA features...\")\n",
    "    mirna_data = harmonized_data['miRNA']\n",
    "    required_mirnas = set(triplets['miRNA'].unique())\n",
    "    available_mirnas = set(mirna_data.index) & required_mirnas\n",
    "    \n",
    "    filtered_mirna = mirna_data.loc[list(available_mirnas)]\n",
    "    filtered_data['miRNA'] = filtered_mirna\n",
    "    print(f\"   miRNA: {len(required_mirnas)} required -> {len(filtered_mirna)} available\")\n",
    "    \n",
    "    # 3. Filter CpG data\n",
    "    print(\"3. Filtering CpG features...\")\n",
    "    cpg_data = harmonized_data['CpG']\n",
    "    \n",
    "    cpg_features_to_keep = []\n",
    "    cpg_rename_mapping = {}\n",
    "    \n",
    "    for cpg_id, actual_cpg in cpg_gene_mapping.items():\n",
    "        if actual_cpg in cpg_data.index:\n",
    "            cpg_features_to_keep.append(actual_cpg)\n",
    "            cpg_rename_mapping[actual_cpg] = cpg_id\n",
    "    \n",
    "    if cpg_features_to_keep:\n",
    "        filtered_cpg = cpg_data.loc[cpg_features_to_keep]\n",
    "        # Rename index\n",
    "        filtered_cpg.index = [cpg_rename_mapping[idx] for idx in filtered_cpg.index]\n",
    "    else:\n",
    "        # If no matching CpG found, create empty DataFrame\n",
    "        filtered_cpg = pd.DataFrame()\n",
    "    \n",
    "    filtered_data['CpG'] = filtered_cpg\n",
    "    print(f\"   CpG: {len(set(triplets['CpG'].unique()))} required -> {len(filtered_cpg)} available\")\n",
    "    \n",
    "    return filtered_data\n",
    "\n",
    "def prepare_clinical_covariates(clinical_data, common_samples):\n",
    "    \"\"\"Prepare clinical covariates\"\"\"\n",
    "    print(\"\\nPreparing covariate data...\")\n",
    "    \n",
    "    # Create sample ID mapping\n",
    "    clinical_sample_mapping = {}\n",
    "    for _, row in clinical_data.iterrows():\n",
    "        base_id = f\"P{row['NTUID']}\"\n",
    "        clinical_sample_mapping[base_id] = row\n",
    "    \n",
    "    # Build covariate matrix\n",
    "    covariate_data = []\n",
    "    \n",
    "    for sample in common_samples:\n",
    "        if sample in clinical_sample_mapping:\n",
    "            row_data = clinical_sample_mapping[sample]\n",
    "            covariate_data.append({\n",
    "                'sample_id': sample,\n",
    "                'age': row_data.get('age', np.nan),\n",
    "                'BMI': row_data.get('BMI', np.nan),\n",
    "                'B12_status': row_data.get('B12_status', 'Unknown'),\n",
    "                'parity': row_data.get('parity', np.nan),\n",
    "                'batch_miRNA': row_data.get('batch_miRNA', 1),\n",
    "                'batch_mRNA': row_data.get('batch_mRNA', 1)\n",
    "            })\n",
    "    \n",
    "    covariate_df = pd.DataFrame(covariate_data).set_index('sample_id')\n",
    "    \n",
    "    # Handle categorical variables\n",
    "    if 'B12_status' in covariate_df.columns:\n",
    "        b12_dummies = pd.get_dummies(covariate_df['B12_status'], prefix='B12', drop_first=True)\n",
    "        covariate_df = pd.concat([covariate_df.drop('B12_status', axis=1), b12_dummies], axis=1)\n",
    "    \n",
    "    # Remove columns with all NaN values\n",
    "    covariate_df = covariate_df.dropna(axis=1, how='all')\n",
    "    \n",
    "    print(f\"   Covariate matrix: {covariate_df.shape}\")\n",
    "    print(f\"   Covariate columns: {list(covariate_df.columns)}\")\n",
    "    \n",
    "    return covariate_df\n",
    "\n",
    "def standardize_expression_data(data, method='log2_zscore'):\n",
    "    \"\"\"Standardize expression data\"\"\"\n",
    "    print(f\"\\nStandardizing data using {method} method...\")\n",
    "    \n",
    "    if method == 'log2_zscore':\n",
    "        # log2(TPM + 1) + Z-score\n",
    "        log_data = np.log2(data + 1)\n",
    "        standardized = pd.DataFrame(\n",
    "            stats.zscore(log_data, axis=1, nan_policy='omit'),\n",
    "            index=data.index,\n",
    "            columns=data.columns\n",
    "        )\n",
    "    elif method == 'zscore_only':\n",
    "        # Z-score only (suitable for CpG data)\n",
    "        standardized = pd.DataFrame(\n",
    "            stats.zscore(data, axis=1, nan_policy='omit'),\n",
    "            index=data.index,\n",
    "            columns=data.columns\n",
    "        )\n",
    "    else:\n",
    "        standardized = data.copy()\n",
    "    \n",
    "    return standardized\n",
    "\n",
    "def adjust_for_covariates(expression_data, covariate_df):\n",
    "    \"\"\"Covariate adjustment\"\"\"\n",
    "    print(f\"Adjusting {len(expression_data)} features for covariates...\")\n",
    "    \n",
    "    if covariate_df.empty or covariate_df.shape[1] == 0:\n",
    "        print(\"   No available covariates, skipping adjustment\")\n",
    "        return expression_data\n",
    "    \n",
    "    # Ensure sample matching\n",
    "    common_samples = list(set(expression_data.columns) & set(covariate_df.index))\n",
    "    if len(common_samples) < 10:\n",
    "        print(\"   Insufficient sample overlap, skipping adjustment\")\n",
    "        return expression_data\n",
    "    \n",
    "    expr_subset = expression_data[common_samples]\n",
    "    cov_subset = covariate_df.loc[common_samples]\n",
    "    \n",
    "    # Remove covariates with NaN values\n",
    "    cov_clean = cov_subset.dropna(axis=1)\n",
    "    if cov_clean.shape[1] == 0:\n",
    "        print(\"   Covariates contain too many missing values, skipping adjustment\")\n",
    "        return expression_data\n",
    "    \n",
    "    adjusted_data = pd.DataFrame(index=expr_subset.index, columns=common_samples)\n",
    "    \n",
    "    for feature in expr_subset.index:\n",
    "        y = expr_subset.loc[feature].values\n",
    "        X = cov_clean.values\n",
    "        \n",
    "        # Remove NaN samples\n",
    "        valid_mask = ~np.isnan(y)\n",
    "        if valid_mask.sum() < 10:\n",
    "            adjusted_data.loc[feature] = y\n",
    "            continue\n",
    "        \n",
    "        try:\n",
    "            reg = LinearRegression().fit(X[valid_mask], y[valid_mask])\n",
    "            residuals = y[valid_mask] - reg.predict(X[valid_mask])\n",
    "            \n",
    "            adjusted_values = np.full_like(y, np.nan)\n",
    "            adjusted_values[valid_mask] = residuals\n",
    "            adjusted_data.loc[feature] = adjusted_values\n",
    "        except:\n",
    "            adjusted_data.loc[feature] = y\n",
    "    \n",
    "    return adjusted_data\n",
    "\n",
    "def save_processed_data(processed_data, output_dir):\n",
    "    \"\"\"Save processed data\"\"\"\n",
    "    print(\"\\n\" + \"=\" * 50)\n",
    "    print(\"SAVING PROCESSED DATA\")\n",
    "    print(\"=\" * 50)\n",
    "    \n",
    "    output_files = {}\n",
    "    \n",
    "    for data_type, data in processed_data.items():\n",
    "        if not data.empty:\n",
    "            filename = f\"3_{data_type}_expr_adjusted.csv\"\n",
    "            filepath = os.path.join(output_dir, filename)\n",
    "            data.to_csv(filepath, encoding='utf-8-sig')\n",
    "            output_files[data_type] = filepath\n",
    "            print(f\"Saved {data_type}: {filepath} ({data.shape})\")\n",
    "        else:\n",
    "            print(f\"Skipped {data_type}: empty data\")\n",
    "    \n",
    "    # Generate processing report\n",
    "    report_path = os.path.join(output_dir, \"3_processing_report.txt\")\n",
    "    with open(report_path, 'w', encoding='utf-8') as f:\n",
    "        f.write(\"Expression Data Standardization Processing Report\\n\")\n",
    "        f.write(\"=\" * 50 + \"\\n\\n\")\n",
    "        \n",
    "        for data_type, data in processed_data.items():\n",
    "            f.write(f\"{data_type} Data:\\n\")\n",
    "            f.write(f\"  Dimensions: {data.shape}\\n\")\n",
    "            f.write(f\"  Features: {len(data)}\\n\")\n",
    "            f.write(f\"  Samples: {len(data.columns) if len(data) > 0 else 0}\\n\")\n",
    "            if len(data) > 0:\n",
    "                f.write(f\"  Sample IDs: {list(data.columns[:5])}\\n\")\n",
    "            f.write(\"\\n\")\n",
    "    \n",
    "    print(f\"Processing report: {report_path}\")\n",
    "    return output_files\n",
    "\n",
    "def main():\n",
    "    \"\"\"Main processing pipeline\"\"\"\n",
    "    print(\"=\" * 80)\n",
    "    print(\"EXPRESSION DATA STANDARDIZATION AND COVARIATE ADJUSTMENT\")\n",
    "    print(\"=\" * 80)\n",
    "    \n",
    "    # 1. Load data\n",
    "    mrna_data, mirna_data, cpg_data, clinical_data, triplets, dna_mrna_integration = load_and_prepare_data()\n",
    "    \n",
    "    # 2. Create mappings\n",
    "    sample_mapping = standardize_sample_ids()\n",
    "    gene_mapping = create_gene_id_mapping(dna_mrna_integration)\n",
    "    \n",
    "    # 3. Harmonize data format\n",
    "    data_dict = {\n",
    "        'mRNA': mrna_data,\n",
    "        'miRNA': mirna_data,\n",
    "        'CpG': cpg_data\n",
    "    }\n",
    "    \n",
    "    harmonized_data = harmonize_sample_names(data_dict, sample_mapping)\n",
    "    \n",
    "    # 4. Get common samples\n",
    "    common_samples = None\n",
    "    for data_type, data in harmonized_data.items():\n",
    "        if common_samples is None:\n",
    "            common_samples = set(data.columns)\n",
    "        else:\n",
    "            common_samples &= set(data.columns)\n",
    "    \n",
    "    common_samples = list(common_samples)\n",
    "    print(f\"\\nCommon sample count: {len(common_samples)}\")\n",
    "    \n",
    "    # 5. Create CpG mapping\n",
    "    cpg_gene_mapping = create_cpg_gene_mapping(cpg_data, triplets)\n",
    "    \n",
    "    # 6. Filter relevant features\n",
    "    filtered_data = filter_triplet_features(harmonized_data, triplets, gene_mapping, cpg_gene_mapping)\n",
    "    \n",
    "    # 7. Prepare covariates\n",
    "    covariate_df = prepare_clinical_covariates(clinical_data, common_samples)\n",
    "    \n",
    "    # 8. Standardize and adjust\n",
    "    processed_data = {}\n",
    "    \n",
    "    for data_type, data in filtered_data.items():\n",
    "        if data.empty:\n",
    "            processed_data[data_type] = data\n",
    "            continue\n",
    "        \n",
    "        print(f\"\\nProcessing {data_type} data...\")\n",
    "        \n",
    "        # Ensure using common samples\n",
    "        data_common = data[common_samples] if len(data) > 0 else data\n",
    "        \n",
    "        # Standardization\n",
    "        if data_type in ['mRNA', 'miRNA']:\n",
    "            standardized = standardize_expression_data(data_common, 'log2_zscore')\n",
    "        else:  # CpG\n",
    "            standardized = standardize_expression_data(data_common, 'zscore_only')\n",
    "        \n",
    "        # Covariate adjustment\n",
    "        if data_type in ['mRNA', 'miRNA']:  # CpG usually not adjusted for covariates\n",
    "            adjusted = adjust_for_covariates(standardized, covariate_df)\n",
    "        else:\n",
    "            adjusted = standardized\n",
    "        \n",
    "        processed_data[data_type] = adjusted\n",
    "    \n",
    "    # 9. Save results\n",
    "    output_files = save_processed_data(processed_data, processed_data_dir)\n",
    "    \n",
    "    print(\"\\n\" + \"=\" * 80)\n",
    "    print(\"PROCESSING COMPLETE!\")\n",
    "    print(\"Output files:\")\n",
    "    for data_type, filepath in output_files.items():\n",
    "        print(f\"  {data_type}: {filepath}\")\n",
    "    print(\"=\" * 80)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1976a096-645c-403d-bfd7-deaa24c8e2ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>> mRNA Expression (first 5 rows):\n",
      "              P87       P42      P131      P204       P64  ...       P68       P86      P191      P207      P246\n",
      "PLEKHG4  0.142834  0.000198  0.895434 -0.583812 -1.013475  ... -1.661380 -0.581513 -0.907411 -0.385621  0.753601\n",
      "SCML2    1.086286  1.113681 -0.020616  0.871553 -0.042929  ...  0.462531  0.029497  0.441360 -0.411833  0.112399\n",
      "LEP     -0.019547 -0.615577 -0.089677  2.017611  0.454004  ...  1.354351  0.397417 -0.284881 -0.355310 -0.704216\n",
      "LHX3     3.152699 -0.711416 -0.374365 -0.462718 -0.141049  ... -0.301088  0.217298  0.006024 -0.426089  2.896053\n",
      "FBXO17  -0.559441 -0.227297  1.382265 -0.247219 -1.358195  ...  0.146625 -0.144769 -0.952598  0.192881 -0.481858\n",
      "\n",
      "[5 rows x 49 columns] \n",
      "\n",
      ">>> miRNA Expression (first 5 rows):\n",
      "                      P87       P42      P131      P204       P64  ...       P68       P86      P191      P207  \\\n",
      "sRNA                                                               ...                                           \n",
      "hsa-miR-18b-5p  -0.802694 -0.833266  1.301456 -0.461765 -1.052004  ... -0.866896 -0.400184 -1.382361 -0.655052   \n",
      "hsa-miR-3679-3p -0.426132  0.725406  0.191884  0.147866 -0.803396  ...  0.593936  0.746279  1.457265  2.302154   \n",
      "hsa-miR-214-3p   1.120534 -0.536941  0.913321 -1.132919 -1.283905  ... -1.071938 -0.536878  0.380677  0.484688   \n",
      "hsa-miR-151b     0.695341 -0.517554  0.191653 -1.462746 -1.261387  ... -1.305989 -1.633482  0.476590  0.779340   \n",
      "hsa-miR-134-3p   1.742487 -0.104930  0.391950 -0.064654 -1.420289  ... -0.181519  1.052676  1.729319  0.332412   \n",
      "\n",
      "                     P246  \n",
      "sRNA                       \n",
      "hsa-miR-18b-5p   0.980686  \n",
      "hsa-miR-3679-3p -0.659228  \n",
      "hsa-miR-214-3p  -0.423809  \n",
      "hsa-miR-151b     0.581010  \n",
      "hsa-miR-134-3p  -0.705156  \n",
      "\n",
      "[5 rows x 49 columns] \n",
      "\n",
      ">>> CpG Expression (first 5 rows):\n",
      "                  P87       P42      P131      P204       P64  ...       P68       P86      P191      P207      P246\n",
      "CpG_PLEKHG4 -1.812731  1.245882  1.333021  0.049053 -0.118637  ... -0.880522  1.547492  1.254034  0.268107 -0.392342\n",
      "CpG_SCML2   -0.900058 -1.509649  0.791163 -0.369555  1.017099  ...  1.004935 -1.505583 -0.587335  1.017030  0.744957\n",
      "CpG_LEP      0.030388  0.007683  0.370955  0.207700  1.101983  ...  0.707110 -1.042590 -1.223364  0.224254  0.604705\n",
      "CpG_LHX3    -0.253595  1.800199  2.283176  0.454108 -0.688832  ...  0.857556 -0.246087 -0.108058  0.012784  0.393584\n",
      "CpG_FBXO17  -0.260077 -0.391493 -1.195651  0.188391  1.313591  ... -0.589825 -0.161880  0.793048  0.064876 -0.259493\n",
      "\n",
      "[5 rows x 49 columns] \n",
      "\n",
      ">>> Triplet List (first 5 rows):\n",
      "           CpG           miRNA     mRNA\n",
      "0  CpG_PLEKHG4  hsa-miR-27a-5p  PLEKHG4\n",
      "1  CpG_PLEKHG4  hsa-miR-365...  PLEKHG4\n",
      "2  CpG_PLEKHG4   hsa-miR-93-3p  PLEKHG4\n",
      "3    CpG_SCML2   hsa-miR-1269a    SCML2\n",
      "4    CpG_SCML2  hsa-miR-18b-5p    SCML2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Define paths\n",
    "base_path = \"/Users/heweilin/Desktop/P056_Code_2/Processed_Data\"\n",
    "path_mrna   = os.path.join(base_path, \"3_mRNA_expr_adjusted.csv\")\n",
    "path_mirna  = os.path.join(base_path, \"3_miRNA_expr_adjusted.csv\")\n",
    "path_cpg    = os.path.join(base_path, \"3_CpG_expr_adjusted.csv\")\n",
    "path_triplet = os.path.join(base_path, \"3_candidate_triplets.csv\")\n",
    "\n",
    "# Load data\n",
    "df_mrna = pd.read_csv(path_mrna, index_col=0)\n",
    "df_mirna = pd.read_csv(path_mirna, index_col=0)\n",
    "df_cpg = pd.read_csv(path_cpg, index_col=0)\n",
    "df_triplets = pd.read_csv(path_triplet)\n",
    "\n",
    "# Display first 5 rows of each\n",
    "print(\">>> mRNA Expression (first 5 rows):\")\n",
    "print(df_mrna.head(), \"\\n\")\n",
    "\n",
    "print(\">>> miRNA Expression (first 5 rows):\")\n",
    "print(df_mirna.head(), \"\\n\")\n",
    "\n",
    "print(\">>> CpG Expression (first 5 rows):\")\n",
    "print(df_cpg.head(), \"\\n\")\n",
    "\n",
    "print(\">>> Triplet List (first 5 rows):\")\n",
    "print(df_triplets.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73107c6d-4ed9-43d3-88aa-75fa6bba8852",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:text_analytics_coursework]",
   "language": "python",
   "name": "conda-env-text_analytics_coursework-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
